storage/src/main/java/org/apache/kafka/server/log/internals/AbstractIndex.java

```diff
@@ -484,27 +478,35 @@ private static MappedByteBuffer createMappedBuffer(RandomAccessFile raf, boolean
     }
 
     /**
-     * Lookup lower and upper bounds for the given target.
+     * Lookup lower or upper bounds for the given target.
      */
-    private BinarySearchResult indexSlotRangeFor(ByteBuffer idx, long target, IndexSearchType searchEntity) {
+    private int indexSlotRangeFor(ByteBuffer idx, long target, IndexSearchType searchEntity,
+                                  SearchType searchType) {
         // check if the index is empty
         if (entries == 0)
-            return new BinarySearchResult(-1, -1);
+            return -1;
 
         int firstHotEntry = Math.max(0, entries - 1 - warmEntries());
         // check if the target offset is in the warm section of the index
         if (compareIndexEntry(parseEntry(idx, firstHotEntry), target, searchEntity) < 0) {
-            return binarySearch(idx, target, searchEntity, firstHotEntry, entries - 1);
+            return binarySearch(idx, target, searchEntity, searchType, firstHotEntry, entries - 1);
         }
 
         // check if the target offset is smaller than the least offset
-        if (compareIndexEntry(parseEntry(idx, 0), target, searchEntity) > 0)
-            return new BinarySearchResult(-1, 0);
+        if (compareIndexEntry(parseEntry(idx, 0), target, searchEntity) > 0) {
+            switch (searchType) {
+                case SMALLEST_UPPER_BOUND:
+                    return -1;
```

Comment:
It seems to me `SMALLEST_UPPER_BOUND` should return 0 and `LARGEST_LOWER_BOUND` should return -1

Reply:
Good catch, I reversed it here and in the other place. Will fix both.